ios/RNMessageStream.m

```diff
@@ -40,14 +42,55 @@ - (instancetype)initWithDisplayInAppNotifications:(BOOL)displayNotifications {
 }
 
 - (BOOL)shouldPresentInAppNotificationForMessage:(MARMessage *)message {
-    NSMutableDictionary *payload = [NSMutableDictionary dictionaryWithDictionary:[message dictionary]];
+    if (self.defaultInAppNotification) {
+        return YES;
+    }
+    __block BOOL result = YES;
+
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        result = [self emitWithTimeout:message];
+
+        dispatch_semaphore_signal(semaphore);
+    });
 
-    if ([message attributes]) {
-        [payload setObject:[message attributes] forKey:@"attributes"];
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+
+    return result;
+}
+
+- (BOOL)emitWithTimeout:(MARMessage *)message {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        NSMutableDictionary *payload = [NSMutableDictionary dictionaryWithDictionary:[message dictionary]];
+
+        if ([message attributes]) {
+            [payload setObject:[message attributes] forKey:@"attributes"];
+        }
+        [self sendEventWithName:@"inappnotification" body:payload];
+        
+        @synchronized (self) {
+            self.inAppNotificationHandled = NO;
+        }
+        dispatch_semaphore_signal(self.eventSemaphore);
```

Comment:
We should remover this signal call; the semaphore should either be signalled by the `notifyInAppHandled` method or it should timeout and continue.